src/utp_stream.cpp

```diff
@@ -2468,8 +2505,32 @@ void utp_socket_impl::ack_packet(packet* p, time_point const& receive_time
 
 	m_rtt.add_sample(rtt / 1000);
 	if (rtt < min_rtt) min_rtt = rtt;
-	// tune receive buffer size according to the new RTT value
-	m_receive_buffer_capacity = adjust_receive_buf_size(m_rtt.mean());
+
+	// update the all-time mininimum RTT value
+	if (m_min_rtt < min_rtt)
+	{
+		m_min_rtt = min_rtt;
+		// and tune receive buffer size according to it
+		boost::int32_t const new_capacity = get_optimized_receive_buf_capacity(m_min_rtt/1000);
+		// m_min_rtt should never increase, but the buffer size could be
+		// increased from its default value after the first RTT. The
+		// evolution of buffer size in accord with min_rtt happens like
+		// this:                       +------------------------------+-----------+----------------+
+		//                             | start                        | first RTT | following RTTs |
+		// +---------------------------+------------------------------+-----------+----------------+
+		// | min_rtt                   | std::numeric_limits<>::max() | decrease  | decrease       |
+		// +---------------------------+------------------------------+-----------+----------------+
+		// | m_receive_buffer_capacity | 1 MByte                      | increase  | decrease       |
+		// +---------------------------+------------------------------+-----------+----------------+
+		// We only use surplus mechanism if the capacity should really
+		// decrease. In case it should increase (e.g. after the first RTT),
+		// we change it directly.
+		if (new_capacity <= m_receive_buffer_capacity - m_receive_buffer_surplus)
+			m_receive_buffer_surplus += (m_receive_buffer_capacity - m_receive_buffer_surplus - new_capacity);
+		else
+			m_receive_buffer_capacity = new_capacity;
```

Comment:
if ``new_capacity`` is 1000, ``m_receive_buffer_capacity`` is 1100 and ``m_receive_buffer_surplus`` is 200. i.e. we want ``m_receive_buffer_capacity`` to be 900, we're just waiting for the right moment to adjust it. This logic doesn't look right:

``new_capacity <= m_receive_buffer_capacity - m_receive_buffer_surplus``, i.e.:

(1000 <= 1100 - 200) -> false

which means we'll set ``m_receive_buffer_capacity`` to 1000 (i.e. not deferring the decrement), and perhaps more importantly, without adjusting the surplus, so now we effectively are looking to decrement the capacity to 800.

I think the second case (the else-block) here should have these asserts:

``TORRENT_ASSERT(m_receive_buffer_capacity < new_capacity);``
``TORRENT_ASSERT(m_receive_buffer_surplus == 0);``

and I think the first case (the if-block) should have this assert after the adjustment has been made:

``TORRENT_ASSERT(m_receive_buffer_surplus >= 0);``

Reply:
(1000 <= 1100 - 200) -> false
case really cannot happen with the current implementation of `get_optimized_receive_buf_capacity()`, because the latter would never recommend buffer size lower than its starting value (that is 1000 in this example). So, if everything goes as planned, the asserts would never trigger.

However, I agree that making pieces interdependent on each other is a bad design pattern. I will redo this procedure to be safe and will add the proposed asserts.

Reply:
I see. yeah, encoding knowledge about the internals of that function there is risky. Imagine a future where the *minimum* RTT is changed to a running average RTT for instance. (I can think of other reasons that's a bad idea, but still)

Reply:
By coincidence, this is almost exactly what I planned to implement today. Real-life test runs of Tribler showed that connection circuit could switch intermediate hops several times during its lifetime, resulting in dramatic changes in connection properties (RTT, bandwidth). Therefore, to support this kind of mutations, we have to use something like a circular buffer of lowest seen RTTs in last 1 minute. Maybe `timestamp_history` object could be modified to fit this purpose.